[WPE] WPE Platform: improve cursor theme detection

carlosgcampos opened the pull request:
#### 86cd382552e0adbe02ecdc24893b5cca26854f68
<pre>
[WPE] WPE Platform: improve cursor theme detection
<a href="https://bugs.webkit.org/show_bug.cgi?id=294745">https://bugs.webkit.org/show_bug.cgi?id=294745</a>

Reviewed by Adrian Perez de Castro.

We iterate paths checking if icons dir is present, but we should
actually try to load the theme and ensure there&apos;s a cursors directory.
Also try with Adwaita cursor theme as a fallback if everything else
failed.

Canonical link: <a href="https://commits.webkit.org/296502@main">https://commits.webkit.org/296502@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed1158751be2e443e3d2a57e5157a945a0cce778

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108753 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28414 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18838 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113964 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59137 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29103 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36978 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82627 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111701 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23123 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97956 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63064 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22544 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58665 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92493 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16142 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117085 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35807 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26438 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91647 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36180 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94223 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91453 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36352 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14105 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31690 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17555 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35707 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41241 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35417 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38764 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37095 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebKit/WPEPlatform/wpe/WPECursorTheme.cpp:
```diff
@@ -73,42 +73,54 @@ static GUniquePtr<char> cursorsPath(const char* basePath, Vector<GUniquePtr<char
+    if (g_strcmp0(name, "default")) {
+        if (auto theme = CursorTheme::create("default", size))
+            return theme;
+    }
```
I might be miss-reading this, but in an scenario where no cursors are present, we would attempt the Adwaita and print the `g_warning` below twice, wouldn't we?

In line 114, we would recursively call  back here attempting to create a `default` theme, it would go to line 113, skip, continue into fallback to Adwaita in line 118, fail, then print warning in 123, return `nullptr`. The caller would, in line 114, continue down as well, fallback again to Adwaita.

So we probably should not try _Adwaita_ if we're attempting to find the `default` theme.

```suggestion
    // If not the default, try with the default.
    if (g_strcmp0(name, "default")) {
        if (auto theme = CursorTheme::create("default", size))
            return theme;
    } else
        return nullptr
```

the above should be able too continue into Adwaita and print a single warning if `theme` is null in line 115.

carlosgcampos replied:
Right, good point!

carlosgcampos replied:
But we should return early when not the default, I'll update the patch.

spenap commented on Source/WebKit/WPEPlatform/wpe/WPECursorTheme.cpp:
```diff
@@ -73,42 +73,54 @@ static GUniquePtr<char> cursorsPath(const char* basePath, Vector<GUniquePtr<char
+    return makeUnique<CursorTheme>(WTFMove(path), size, WTFMove(inheritedThemes));
+}
+
+std::unique_ptr<CursorTheme> CursorTheme::create(const char* name, uint32_t size)
```
I feel like the intent in this method is a loop:

1. user-data {theme} icons
2. home {theme} .icons
3. system-data {theme} icons
4. again, with "default" as theme
5. Adwaita
6. warn and end

I believe the code could be cleaner if we wrote it in that way: looping through a list of paths in priority order, and inside the body of the loop building the filename and doing `tryCreateTheme`

carlosgcampos replied:
You are right, but I tried to avoid having to build all the paths and then iterate them.

aperezdc commented on Source/WebKit/WPEPlatform/wpe/WPECursorTheme.cpp:
```diff
@@ -73,42 +73,57 @@ static GUniquePtr<char> cursorsPath(const char* basePath, Vector<GUniquePtr<char
-    } else
-        g_warning("Could not find any base paths for cursors search");
+    // Fallback to Adwaita.
+    if (auto theme = tryCreateTheme("/usr/share/icons/Adwaita/", size))
```
The `Adwaita` theme could be installed elsewhere in any of the paths returned by `g_get_system_data_dirs()`, so I think it would make more sense to try here with `CursorTheme::create()` in the same way as done for the `default` one, instead of hardcoding the `/usr/share/icons/Adwaita` path:

```suggestion
    if (auto theme = CursorTheme::create("Adwaita", size))
```

carlosgcampos replied:
Good point, I'll rewrite to iterate a loop of theme names, avoiding also the recursive calls when falling back.

aperezdc reviewed (approved):
LGTM overall, but I agree with @spenap that having a loop over the candidate theme names would be clearer. Also, I see no reason to hardcode the path to the `Adwaita` theme&mdash;it should be looked up using the same logic as for all the other lookups.

carlosgcampos committed:
[WPE] WPE Platform: improve cursor theme detection
https://bugs.webkit.org/show_bug.cgi?id=294745

Reviewed by Adrian Perez de Castro.

We iterate paths checking if icons dir is present, but we should
actually try to load the theme and ensure there's a cursors directory.
Also try with Adwaita cursor theme as a fallback if everything else
failed.

Canonical link: https://commits.webkit.org/296502@main
